itest/lnd_sendonion_test.go

```diff
@@ -0,0 +1,143 @@
+	alice, bob := ht.Alice, ht.Bob
+	carol := ht.NewNode("carol", nil)
+	dave := ht.NewNode("dave", nil)
+
+	// Connect nodes to ensure propagation of channels.
+	ht.EnsureConnected(alice, bob)
+	ht.EnsureConnected(bob, carol)
+	ht.EnsureConnected(carol, dave)
+
+	const chanAmt = btcutil.Amount(100000)
+
+	// We'll open a channel between Alice and Bob with Alice's liquidity
+	// drained (eg: all funds on Bob's side of the channel) to exercise the
+	// RPC logic which selects the appropriate channel link.
+	chanPointBobAlice := ht.OpenChannel(
+		bob, alice, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(alice, chanPointBobAlice)
+
+	// Now, open a channel with 100k satoshis between Alice and Bob with
+	// Alice being the sole funder of the channel. This will provide Alice
+	// with outbound liquidity she can use to complete payments.
+	chanPointAliceBob := ht.OpenChannel(
+		alice, bob, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(alice, chanPointAliceBob)
+
+	// We'll create Dave and establish a channel between Bob and Carol.
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, dave)
+	chanPointBob := ht.OpenChannel(
+		bob, carol, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(bob, chanPointBob)
+
+	// Next, we'll create Carol and establish a channel to from her to
+	// Dave.
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, carol)
+	chanPointCarol := ht.OpenChannel(
+		carol, dave, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(carol, chanPointCarol)
+
+	// Make sure Alice knows the channel between Bob and Carol.
+	ht.AssertTopologyChannelOpen(alice, chanPointBob)
+	ht.AssertTopologyChannelOpen(alice, chanPointCarol)
+
+	const (
+		numPayments = 1
+		paymentAmt  = 10000
+	)
+
+	// Request an invoice from Dave so he is expecting payment.
+	_, rHashes, invoices := ht.CreatePayReqs(dave, paymentAmt, numPayments)
+	var preimage lntypes.Preimage
+	copy(preimage[:], invoices[0].RPreimage)
+
+	// Query for routes to pay from Alice to Dave.
+	routesReq := &lnrpc.QueryRoutesRequest{
+		PubKey: dave.PubKeyStr,
+		Amt:    paymentAmt,
+		// AmtMsat: paymentAmt,
+	}
+	routes := alice.RPC.QueryRoutes(routesReq)
+	route := routes.Routes[0]
+	finalHop := route.Hops[len(route.Hops)-1]
+	finalHop.MppRecord = &lnrpc.MPPRecord{
+		PaymentAddr:  invoices[0].PaymentAddr,
+		TotalAmtMsat: int64(lnwire.NewMSatFromSatoshis(paymentAmt)),
+	}
+
+	// Construct an onion for the route from Alice to Dave.
+	paymentHash := rHashes[0]
+	onionReq := &routerrpc.BuildOnionRequest{
+		Route:       route,
+		PaymentHash: paymentHash,
+	}
+	onionResp := alice.RPC.BuildOnion(onionReq)
+
+	// Dispatch a payment via the SendOnion RPC.
+	firstHop := bob.PubKey
+	sendReq := &routerrpc.SendOnionRequest{
+		FirstHopPubkey: firstHop[:],
+		Amount:         route.TotalAmtMsat,
+		Timelock:       route.TotalTimeLock,
+		PaymentHash:    paymentHash,
+		OnionBlob:      onionResp.OnionBlob,
+		AttemptId:      0,
```

Comment:
I propose to use some other value just to check that it is passed correctly (0 is the default for integer types).